### ISSUE TITLE
[ERROR] Unable to discover poll endpoint: InvalidSignatureException: Signature not yet current

### ISSUE THREAD

**Issue by bploetz** [aws/amazon-ecs-agent#1513]
### Summary
Old tasks stuck in RUNNING after deploy.

### Description
Just deployed a new version of one of our images to our cluster of 48 instances. 42 of them deployed fine, but 6 are stuck in this weird state where the old task definition is still running (LAST STATUS is RUNNING) but it's desired state is STOPPED.

I then noticed that the last 24 hours worth ECS agent logs were filled with this error:

```
2018-08-09T19:09:01Z [ERROR] Unable to discover poll endpoint: InvalidSignatureException: Signature not yet current: 20
180809T190901Z is still later than 20180809T190834Z (20180809T190334Z + 5 min.)
	status code: 400, request id: ea3ec93c-9c06-11e8-bd0a-45b58ea5e96f
2018-08-09T19:09:01Z [INFO] Error from tcs; backing off: InvalidSignatureException: Signature not yet current: 20180809
T190901Z is still later than 20180809T190834Z (20180809T190334Z + 5 min.)
	status code: 400, request id: ea3ec93c-9c06-11e8-bd0a-45b58ea5e96f
2018-08-09T19:09:06Z [ERROR] Unable to discover poll endpoint, err: InvalidSignatureException: Signature not yet current: 20180809T190906Z is still later than 20180809T190839Z (20180809T190339Z + 5 min.)
	status code: 400, request id: ed2dd297-9c06-11e8-bd0a-45b58ea5e96f
2018-08-09T19:09:06Z [INFO] Reconnecting to ACS in: 2m5.113640085s
```

I tried running command with the aws cli (as noted in #676, another unresolved ticket about this same error):

```
aws --region us-east-1  ecs discover-poll-endpoint --cluster XXXXXX --container-instance arn:aws:ecs:us-east-1:XXXXXXXX:container-instance/XXXXXXXXX

An error occurred (InvalidSignatureException) when calling the DiscoverPollEndpoint operation: Signature not yet current: 20180809T200315Z is still later than 20180809T200249Z (20180809T195749Z + 5 min.)
```

### Expected Behavior
All containers should have rolled to the new task definition version.

### Observed Behavior
All but 6 containers rolled to the new task definition version.

### Environment Details
docker info
```
docker info
Containers: 2
 Running: 2
 Paused: 0
 Stopped: 0
Images: 3
Server Version: 17.09.1-ce
Storage Driver: devicemapper
 Pool Name: docker-docker--pool
 Pool Blocksize: 524.3kB
 Base Device Size: 10.74GB
 Backing Filesystem: ext4
 Data file: 
 Metadata file: 
 Data Space Used: 1.001GB
 Data Space Total: 23.33GB
 Data Space Available: 22.33GB
 Metadata Space Used: 442.4kB
 Metadata Space Total: 25.17MB
 Metadata Space Available: 24.72MB
 Thin Pool Minimum Free Space: 2.333GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.76-3.78.amzn1.x86_64
Operating System: Amazon Linux AMI 2017.09
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 993.3MiB
Name: ip-10-0-46-66
ID: XZAS:UA6V:XOWM:7V6D:HLU3:QB23:RMGP:QTFS:JIYN:JR4C:6HSQ:MMRH
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

curl http://localhost:51678/v1/metadata
```
curl http://localhost:51678/v1/metadata
{"Cluster":"XXXXXXXXXX","ContainerInstanceArn":"arn:aws:ecs:XXXXXX:XXXXXXXXX:container-instance/XXXXXXXXXX","Version":"Amazon ECS Agent - v1.16.2 (998c9b5)"}
```

I'm going to manually kill all but one of these bad instances in case we need it for debugging.

**Comment by bploetz** [comment]
The lone remaining instance exhibiting the behavior is i-0e280020dbcad660d

**Comment by adnxn** [comment]
@bploetz: This may need deeper investigation, but my naive suspicion is that it might be a clock skew issue with the ec2 instance or docker which is affecting the agent container. Will follow up with more information as we investigate further.

**Comment by bploetz** [comment]
FWIW, I had this happen to two more instances on my next deploy. Stopping the zombie task and the ecs agent did nothing. I then tried removing the ecs agent checkpoint file and bringing up the agent from scratch to no avail (the same error spewed over and over in the logs). I ended up nuking the EC2 instances. i-0e280020dbcad660d is still alive if you need one to debug.

Not sure if it matters or not, but these are spot pool instances.

**Comment by adnxn** [comment]
@bploetz: The error seems unrelated to #676. The error you're seeing is caused by `InvalidSignatureException: Signature not yet current...` and that's resulting in failures to discover the endpoint. 

Have you seen this [post on clock skew](https://serverfault.com/questions/724114/attempting-to-initialize-aws-eb-getting-signature-not-yet-current-error)? If you have the instance around - Would you be able to try their suggestion and try `aws --region us-east-1  ecs discover-poll-endpoint` again?

Let us know if that unblocks you for making requests using the cli.

**Comment by bploetz** [comment]
It looks like I may be hitting this issue, as the instances effected are t2.micro instances

https://forums.aws.amazon.com/thread.jspa?messageID=819826&tstart=0

I removed NTP and installed Chrony (as noted here https://aws.amazon.com/blogs/aws/keeping-time-with-amazon-time-sync-service/) and once I started the Chrony service, the ECS agent was able to connect to the discovery endpoint again.

```
2018-08-22T13:30:33Z [INFO] Error from tcs; backing off: InvalidSignatureException: Signature expired: 20180822T133033Z is now earlier than 20180822T133135Z (20180822T133635Z - 5 min.)
	status code: 400, request id: 6347d700-a610-11e8-a319-f3efbf3791e9
2018-08-22T13:30:46Z [ERROR] Unable to discover poll endpoint, err: InvalidSignatureException: Signature expired: 20180822T133046Z is now earlier than 20180822T133148Z (20180822T133648Z - 5 min.)
	status code: 400, request id: 6b094ff2-a610-11e8-a319-f3efbf3791e9
2018-08-22T13:30:46Z [INFO] Reconnecting to ACS in: 2m7.008720675s
2018-08-22T13:37:36Z [INFO] Connected to TCS endpoint
2018-08-22T13:38:55Z [INFO] Connected to ACS endpoint
```

Looks like this isn't an ecs agent issue after all. Thanks for your help guys!
